extension/golang/environment.go

```diff
@@ -0,0 +1,420 @@
+				log.Warning("Found golang extension without plugin - ignored")
+				continue
+			}
+			url := strings.TrimPrefix(extension.URL, "file://")
+			err := thisEnvironment.Load(url, code)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// setup time limits for matching extensions
+	thisEnvironment.timeLimit = timeLimit
+	for _, timeLimit := range timeLimits {
+		if timeLimit.Match(path) {
+			thisEnvironment.timeLimits = append(thisEnvironment.timeLimits, schema.NewEventTimeLimit(timeLimit.EventRegex, timeLimit.TimeDuration))
+		}
+	}
+	return nil
+}
+
+// HandleEvent handles an event
+func (thisEnvironment *Environment) HandleEvent(event string, context map[string]interface{}) (err error) {
+	context["event_type"] = event
+	handlers, ok := thisEnvironment.extEnv.Handlers[event]
+
+	if ok {
+		err = thisEnvironment.dispatchEvent(handlers, event, context)
+	}
+	if err != nil {
+		log.Error("dispatchEvent Error: %s", err)
+		return err
+	}
+
+	schemasHandlersList, ok := thisEnvironment.extEnv.HandlersSchema[event]
+
+	if ok {
+		for schemaID, handlers := range schemasHandlersList {
+			err = thisEnvironment.dispatchEventWithResource(schemaID, handlers, event, context)
+
+			if err != nil {
+				log.Error("dispatchEventWithResource Error: %s", err)
+				return err
+			}
+		}
+	}
+
+	return
+}
+
+//@todo: this and below function (dispatchEventWithResource) should be somehow unified because they're very similar but operates on different types
+func (thisEnvironment *Environment) dispatchEvent(handlers goext.PriorityHandlerList, event string, context map[string]interface{}) (err error) {
+	//@todo: it's a poor idea to sort the prioritized handlers each time the event is being handled.
+	priorities := []int{}
+	for priority := range handlers {
+		priorities = append(priorities, int(priority))
+	}
+
+	sort.Ints(priorities)
+
+	for priority := range priorities {
+		for _, fn := range handlers[goext.Priority(priority)] {
+			err := fn(context, &thisEnvironment.extEnv)
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return
+}
+
+func (thisEnvironment *Environment) dispatchEventWithResource(schemaID string, handlers goext.HandlerSchemaListOfPriorities, event string, context map[string]interface{}) (err error) {
+	//@todo: it's a poor idea to sort the prioritized handlers each time the event is being handled.
+	priorities := []int{}
+	for priority := range handlers {
+		priorities = append(priorities, int(priority))
+	}
+
+	sort.Ints(priorities)
+
+	resource, err := thisEnvironment.schemaIDToResource(schemaID, context)
+
+	if err != nil {
+		return err
+	}
+
+	for priority := range priorities {
+		for _, fn := range handlers[goext.Priority(priority)] {
+			err := fn(context, resource, &thisEnvironment.extEnv)
+
+			for key, value := range thisEnvironment.resourceToMap(resource).(map[string]interface{}) {
+				if _, ok := context["resource"].(map[string]interface{})[key]; ok {
+					context["resource"].(map[string]interface{})[key] = value
```

Comment:
as far as I can tell, the handlers are expected to use resource directly (instead of getting it from context). two implications:

1. it'll be enough to update context only once, on function exit (think defer)
2. we should make the context easy to use correctly and hard to use incorrectly (prohibit accessing context["resource"]). I'm suggesting changing the context type to some struct instead of being map[string]interface{}